src/Microsoft.Extensions.Configuration.AzureAppConfiguration/FeatureManagement/FeatureFlagOptions.cs

```diff
@@ -28,6 +28,12 @@ public class FeatureFlagOptions
         /// </summary>
         public TimeSpan CacheExpirationInterval { get; set; } = RefreshConstants.DefaultFeatureFlagsCacheExpirationInterval;
 
+        /// <summary>
+        /// Forces feature flags and dynamic features to be written into the configuration system using the  
+        /// feature management V2.0 schema introduced in Microsoft.FeatureManagement 3.0.0-preview library.
+        /// </summary>
+        public bool EnableFeatureManagementV2SchemaPreview { get; set; } = false;
```

Comment:
I suggest us not to introduce a new API for opt-in/opt-out.

Similar to the tracing, let people opt-in/out-out through an environment variable. I would like to avoid for people hard-coding choices in code. We all know how hard it is to make people update their code later.

Reply:
The plan was to only have this API available in 5.x preview release. So it will be impossible for users to avoid updating their code if they move to a stable release.

Tracing is a passive component. Opt-in or opt-out has zero effect on the behavior of the application. I definitely see why we went with environment variable there. I don't know if that carries over to this though. I can't imagine it would be good for user to set the environment variable in 5.2-preview and then move to 5.3 stable and be scratching their heads at why it's not working as expected.

Reply:
I mentioned the tracing for its opt-in/out mechanism, not necessarily for its reasoning.

For previews, people define an environment variable for their apps to opt-in to v2 behavior. For stable release, we will have a new environment variable for people to opt-out v2 (or opt-in to v1) behavior. No code change is required. For people who mess up, we do exceptions with clear guidance on what they need to do.

I understand your desire for a breaking change between preview and stable as a notification. Do you plan to have an API in stable and maintain it forever? I'm more concerned about the stable version.

PS. For testing, I don't mind if we define an interval variable to avoid the environment variable (if that's the problem).

Reply:
We could have an environment variable `FeatureManagementSchema={v1, v2}` and always honor it. The only change between preview and stable is the default value when the variable is not defined. In the preview, it defaults to v1 and v2 for stable.

Reply:
In addition, the breaking change only has an impact on people who opt-in v2 during the preview. Why do we want to break them when they move to stable and v2 becomes the default?

Reply:
> Do you plan to have an API in stable and maintain it forever? 

Would environment variables that control application behavior have less of a maintenance expectation?

> For people who mess up, we do exceptions with clear guidance on what they need to do.

Wanted to make sure I understood you on this one. Do you mean if they are using a 5.x stable release that doesn't respect the opt-in environment variable that we would throw an exception?

> In addition, the breaking change only has an impact on people who opt-in v2 during the preview. Why do we want to break them when they move to stable and v2 becomes the default

I like that point, the environment variable does save us in that regard.

Reply:
Your point about the maintenance of the environment variable is valid. The difference is it lives outside of the application. It's easier for us to evolve it in the long run (for example, when we introduce v3 and it's not a breaking change in the future). Once people hard code a version in code, it's part of the build. It is harder to change. It's much easier for people to change the environment.

Reply:
> Do you mean if they are using a 5.x stable release that doesn't respect the opt-in environment variable that we would throw an exception?

No, we should always honor the user's choice even between the preview and stable. What I meant is that the stable version is used together with v1 feature flags (but not opt-in to v1).